isogeny_primes.py

```diff
@@ -730,85 +781,140 @@ def get_C_primes(K, G_K, frak_q, epsilons, q_class_group_order, loop_curves=Fals
+    contains_imaginary_quadratic, contains_hilbert_class_field = contains_imaginary_quadratic_field(K)
 
     if contains_hilbert_class_field:
-        if K.absolute_degree() < Kgal.absolute_degree():
-            # It _would_ be nice to remove the Galois restriction in Momose Lemma 1 ...
-            raise NotImplementedError("The Galois closure of the number field you "
-                "entered contains the Hilbert Class field of an imaginary quadratic "
-                "field. Our algorithm currently cannot deal with this instance.")
-        else:
-            raise ValueError("The number field you entered contains the Hilbert "
-                            "Class field of an imaginary quadratic field. The set "
-                            "of isogeny primes in this case is therefore infinite.")
+        raise ValueError("The number field you entered contains the Hilbert "
+                        "Class field of an imaginary quadratic field. The set "
+                        "of isogeny primes in this case is therefore infinite.")
 
-    aux_primes = Kgal.primes_of_bounded_norm(norm_bound)
+    aux_primes = K.primes_of_bounded_norm(norm_bound)
+
+    # First get the emergency aux primes
+
+    C_K = K.class_group()
+    h_K = C_K.order()
+    completely_split_rat_primes = K.completely_split_primes(B=500)
+
+    if contains_imaginary_quadratic:
+
+        good_primes = [p for p in completely_split_rat_primes if gcd(p,6*h_K) == 1]
+        list_of_gens = list(C_K.gens())
+        i = 0
+        while list_of_gens and (i < len(good_primes)):
+            a_good_prime = good_primes[i]
+            emergency_prime_candidates = K.primes_above(a_good_prime)
+
+            for candidate in emergency_prime_candidates:
+                emergency_gen = C_K(candidate)
+                if emergency_gen in list_of_gens:
+                    if a_good_prime > norm_bound:
+                        aux_primes.append(candidate)
+                        logging.debug("Emergency aux prime added: {}".format(candidate))
+                    list_of_gens.remove(emergency_gen)
+            i += 1
+
+        if list_of_gens:
+            raise RuntimeError("We have been unable to add enough emergency "
+                            "auxiliary primes. Try increasing the `B` parameter above.")
+    else:
+        a_good_prime = completely_split_rat_primes[0]
+        candidate = K.primes_above(a_good_prime)[0]
+        if a_good_prime > norm_bound:
+            aux_primes.append(candidate)
+            logging.debug("Emergency aux prime added: {}".format(candidate))
+
+    embeddings = K.embeddings(Kgal)
+    if K.is_galois():
+        G_K = K.galois_group()
+        epsilons = get_pre_type_one_two_epsilons(K.degree(), galgp=G_K)
+    else:
+        epsilons = get_pre_type_one_two_epsilons(K.degree())
+
+    # Do the unit computation first
+
+    unit_gens = K.unit_group().gens_values()
```

Comment:
Maybe give the integers coming from units a name as well (get_U_integers) and put the code for it in a separate function.

Reply:
done.